File: front/src/contractor/model/Contractor.js
import {contacts_api_adapter, createContacts} from "../../contact/model/Contact";

class Contractors extends Array {}

const contractor_api_adapter = contractor => {
    if (contractor["contacts"]) {
        contractor["contacts"] = createContacts(
            contacts_api_adapter(contractor["contacts"])
        );
    }
    return contractor;
};

const contractors_api_adapter = contractors => contractors.map(contractor_api_adapter);

const contractor_view_adapter = contractor => {
    return contractor;
};

const createContractors = (data = []) => {
    const contractors = Contractors.from(data, contractor =>
        createContractor(contractor)
    );
    return contractors;
};

const createContractor = ({
    id = null,
    name = "",
    contractor_type = "",
    contractor_type_label = "",
    address = "",
    phone = "",
    email = "",
    comments = "",
    contract = null, // contract id
    contacts = [],
} = {}) => {
    const publicApi = {
        id,
        name,
        contractor_type,
        contractor_type_label,
        address,
        phone,
        email,
        comments,
        contract,
        contacts,
    };

    return Object.freeze(publicApi);
};

export {
    createContractor as default,
    createContractors,
    contractor_api_adapter,
    contractors_api_adapter,
    contractor_view_adapter,
};

File: back/app/models/contractor.py
from django.db import models

from app.models.contact import Contact
from app.models.contact_relationship import ContractorContact


class Contractor(models.Model):
    class Meta(object):
        db_table = "contractor"
        verbose_name = "Contratista"
        verbose_name_plural = "Contratistas"

    id = models.AutoField(primary_key=True)
    name = models.CharField("Nombre", max_length=100)
    contractor_type = models.CharField("Tipo de contratista", max_length=50, null=True)
    address = models.TextField("Dirección", max_length=500, null=True)
    phone = models.CharField("Celular", max_length=20, null=True)
    email = models.CharField("E-mail", max_length=255, null=True)
    comments = models.TextField("Observaciones", max_length=500, null=True)
    contacts = models.ManyToManyField(Contact, through=ContractorContact)

    def __str__(self):
        return self.name
